File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from _collections import deque
input = sys.stdin.readline

n = int(input().strip())

graph = []

for _ in range(n):
    graph.append(input().strip())
# 0은 미방문 , 1은 방문
visited = [[0]*n for _ in range(n)]

# 단지수 cnt, 단지내 집의수
cnt = 0
ans = []
def bfs(graph,start):
    global visited
    sx,sy = start
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    # 방문체크와 단지내 집 개수 house
    house = 1
    visited[sx][sy] = 1
    q = deque([(sx,sy)])
    while q:
        x,y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= n or ny < 0 or ny >=n or visited[nx][ny] or graph[nx][ny] == '0':continue
            visited[nx][ny] = 1
            house += 1
            q.append((nx,ny))
    return house

for i in range(n):
    for j in range(n):
        if not visited[i][j] and graph[i][j] == '1':
            cnt += 1
            ans.append(bfs(graph,(i,j)))
# 정렬
ans.sort()
print(cnt)
for i in range(len(ans)):
    print(ans[i])

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
//
// Created by 전재현 on 2022/12/29.
//
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <algorithm>
using namespace std;
#define MAX 26

typedef struct {
    int x;
    int y;
}pos;

int n ;
int visited[MAX][MAX] = {0,};
vector <string> graph;



int bfs(pos p){
    queue <pos> q;
    q.push(p);
    //현재 노드 방문체크
    visited[p.x][p.y] = 1;
    int dx[4] = {-1,1,0,0};
    int dy[4] = {0,0,-1,1};
    int x,y,nx,ny;
    int house = 1;
    while ( !q.empty()){
        x = q.front().x;
        y = q.front().y;
        q.pop();
        for (int i = 0 ; i < 4 ; i++){
            nx = x + dx[i];
            ny = y + dy[i];
            // 방문했던경우나 집이없거나 범위벗어나면 패스
            if ( nx < 0 || nx >= n || ny < 0 || ny >= n || visited[nx][ny] == 1 || graph[nx][ny] == '0')continue;
            // 방문체크
            visited[nx][ny] = 1;
            house += 1;
            q.push({nx,ny});
        }
    }
    return house;

}

int main(){
    cin >> n;
    for (int i = 0; i < n; i++){
        string x;
        cin >> x;
        graph.push_back(x);
    }
    int cnt = 0;
    vector <int> ans;
    for (int i = 0; i < n ; i++){
        for (int j = 0 ; j < n ; j++){
            // 방문하지않은곳이며 집이있는곳이면 bfs 수행
            if (visited[i][j] == 0 && graph[i][j] == '1'){
                cnt += 1;
                ans.push_back(bfs({i,j}));
            }
        }
    }
    sort(ans.begin(),ans.end());
    cout << cnt << endl;
    for (auto i = ans.begin(); i < ans.end();++i){
        cout << *i << endl;
    }


}